realhf/api/core/data_api.py

```diff
@@ -1,33 +1,366 @@
 import numpy as np
+import torch
 import torch.utils.data
 import transformers
 
+# NOTE: We only use pandatic dataclasses for SequenceSample
+# such that it will perform automatic checks.
+from pydantic import Field
+from pydantic import dataclasses as pdclasses
+from pydantic import field_validator, model_validator
+
 from realhf.api.core import config as config_api
-from realhf.api.core import model_api
-from realhf.base import datapack, logging, namedarray
+from realhf.base import datapack, logging
 from realhf.base.cluster import spec as cluster_spec
 
 logger = logging.getLogger("api.data")
 
 
+def load_hf_tokenizer(
+    model_name_or_path: str,
+    fast_tokenizer=True,
+    padding_side: Optional[str] = None,
+) -> transformers.PreTrainedTokenizerFast:
+    kwargs = {}
+    if padding_side is not None:
+        kwargs["padding_side"] = padding_side
+    tokenizer = transformers.AutoTokenizer.from_pretrained(
+        model_name_or_path, fast_tokenizer=fast_tokenizer, **kwargs
+    )
+    if tokenizer.pad_token_id is None:
+        tokenizer.pad_token_id = tokenizer.eos_token_id
+    return tokenizer
+
+
+@pdclasses.dataclass
+class SequenceSplitSpec:
+    partitions: Optional[List[Tuple[int, int]]] = None
+    sizes: Optional[List[int]] = None
+
+    @model_validator(mode="after")
+    def _validate_partitions(self) -> "SequenceSplitSpec":
+        if self.partitions is not None:
+            bound = 0
+            for start, end in self.partitions:
+                if start >= end:
+                    raise ValueError(f"Partition {start}-{end} is empty.")
+                if start != bound:
+                    raise ValueError(f"Partition {start}-{end} is not contiguous.")
+                bound = end
+
+        if self.sizes is None and self.partitions is None:
+            raise ValueError("Either sizes or partitions must be provided.")
+        elif self.sizes is not None and self.partitions is not None:
+            if len(self.sizes) != len(self.partitions):
+                raise ValueError("Sizes and partitions are not the consistent.")
+            if self.sizes != [end - start for start, end in self.partitions]:
+                raise ValueError("Sizes and partitions are not the consistent.")
+        elif self.sizes is None:
+            self.sizes = [end - start for start, end in self.partitions]
+        elif self.partitions is None:
+            offsets = np.cumsum([0] + self.sizes)
+            self.partitions = [
+                (offsets[i], offsets[i + 1]) for i in range(len(self.sizes))
+            ]
+
+        return self
+
+
+@pdclasses.dataclass(config=dict(arbitrary_types_allowed=True))
+class SequenceSample:
+    keys: Set[str]
+    trailing_shapes: Dict[str, torch.Size | Tuple]
+    dtypes: Dict[str, torch.dtype]
+
+    ids: List[Hashable]
+
+    seqlens: Dict[str, List[torch.Tensor]]
+
+    data: Optional[Dict[str, torch.Tensor | None]] = None
+
+    metadata: Dict[str, Any] = Field(default_factory=dict)
+
```

Comment:
Maybe we can implement a customized initialization method based on keys to avoid long codes when initializing this class? For example, passes in several (key, trailing_shapes, seqlens, data, metadata) tuples (or dataclasses) + ids and this method could automatically convert them to dict attributes in SequenceSample.